mypy/plugin.py

```diff
@@ -53,6 +54,14 @@ def named_generic_type(self, name: str, args: List[Type]) -> Instance:
         raise NotImplementedError
 
 
+class SemanticAnalyzerPluginInterface:
+    """Interface for accessing semantic analyzer functionality in plugins."""
+
+    @abstractmethod
+    def named_type(self, qualified_name: str, args: Optional[List[Type]] = None) -> Instance:
```

Comment:
Is this the only method that would be potentially used by plugins? Do we need to add some more?

Reply:
My current explorations also need `parse_bool` and `anal_type`

Reply:
`fail` could be useful too. Should I add them as needed or add them now?

Reply:
> Should I add them as needed or add them now?

I am not sure why we actually need this class. Can't we just pass everything? Or am I missing something?

Reply:
I'm following the pattern of the other plugins.  Technically the object that's passed in is the full object. I believe this base class is here to limit what methods the plugins should be calling. i.e. if the method isn't listed here then the mypy checker will complain about it.

Reply:
Then we probably need to think what to add here. My expectation is that many methods of ``SemanticAnalyzer`` may be useful. You can play more with your implementation for ``attrs`` and add here whatever you ever used.

Reply:
I think we can keep only these methods, and then add more in your PR with the actual plugin (if necessary).

Reply:
Here is why we are using an ABC instead exposing the entire semantic analyzer interface:

* By only giving access to a subset of an interface, we declare the public part of the semantic analyzer interface. This has a few benefits related to modularity and loose coupling:
  * Everything else in the semantic analyzer can be more freely modified without breaking plugins. If we change the public interface, we may break 3rd party plugins. (Currently the plugin interface is still not officially supported so we can change things without worries, but I'd like to make the interface more stable in the future.)
  * It's less likely that plugins will rely on internal implementation details that are likely to change, or on poorly thought-out internal APIs that are hard to use correctly.
  * It's easier to reason about plugins in general, as the plugins are expected to stick to the exposed public interface instead of freely accessing mypy internals.
* This can avoid a cyclic import dependency (maybe not right now, but it will help untangle cyclic dependencies). They slow down incremental checking.